network/transport/grpc/connection.go

```diff
@@ -210,22 +206,29 @@ func (mc *conn) registerStream(protocol Protocol, stream Stream) bool {
 	mc.startSending(protocol, stream)
 
 	// A connection can have multiple active streams, but if one of them is closed, all of them should be closed, also closing the underlying connection.
-	go func(cancel func()) {
+	go func() {
 		<-stream.Context().Done()
-		cancel()
-	}(mc.cancelCtx)
+		mc.cancelCtx()
+	}()
 
 	return true
 }
 
 func (mc *conn) startReceiving(protocol Protocol, stream Stream) {
 	peer := mc.Peer() // copy Peer, because it will be nil when logging after disconnecting.
 	atomic.AddInt32(&mc.activeGoroutines, 1)
-	go func(activeGoroutines *int32, cancel func()) {
+	go func(activeGoroutines *int32) {
 		defer atomic.AddInt32(activeGoroutines, -1)
 		for {
 			message := protocol.CreateEnvelope()
 			err := stream.RecvMsg(message)
+			select {
+			case <-mc.ctx.Done():
+				// disconnect: drop message and stop receiving
+				return
+			default:
```

Comment:
interesting pattern, but note that `RecvMsg` is blocking anyway, so you could just check `ctx.Err() != nil`?